src/relay/udprelay/redir_local.rs

```diff
@@ -94,10 +94,16 @@ pub async fn run(context: SharedContext) -> io::Result<()> {
 
     info!("shadowsocks UDP redirect listening on {}", local_addr);
 
-    // NOTE: Associations are only eliminated by expire time
+    // NOTE: Associations are only eliminated by expire time by default
     // So it may exhaust all available file descriptors
     let timeout = context.config().udp_timeout.unwrap_or(DEFAULT_TIMEOUT);
-    let assoc_map: SharedAssocMap = Arc::new(Mutex::new(LruCache::with_expiry_duration(timeout)));
+    let max_conns = context.config().udp_max_conns.unwrap_or(0) as usize;
+
+    let assoc_map: SharedAssocMap = Arc::new(Mutex::new(if max_conns > 0 {
```

Comment:
This is bad, why not

```rust
let max_assoc_opt = context.config().udp_max_associations;
if let Some(ma) = max_assoc_opt {
    LruCache::with_expiry_duration_and_capacity(timeout, ma)
} else {
    LruCache::with_expiry_duration(timeout)
}
```

Reply:
Same in the others.

Reply:
```rust
let assoc_map = if let ...
let assoc_map = Arc::new(Mutex::new(assoc_map));
```

Reply:
If done as per your first comment, then it's possible for a user to configure udp_max_associations = 0. I'm not certain as to the behavior of the cache were that to happen. Adding an additional bounds check for that case inside the if let clause would make the code even longer. Which is why I chose to unwrap_or(0) first and then do the bounds checking. What do you think?

Reply:
`udp_max_associations = 0` should be checked as an error and report it to user.

If user have explictly specify a value for this configure option, you should use it instead of making one of the value to be the same as not configured.